File: 0704/binary_search.py
from typing import *


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        ng, ok = -1, len(nums)
        is_ok = lambda nums, mid, target: nums[mid] >= target
        while abs(ok - ng) > 1:
            mid = int((ok + ng) / 2)
            if is_ok(nums, mid, target):
                ok = mid
            else:
                ng = mid
        return ok if ok < len(nums) and nums[ok] == target else -1


if __name__ == '__main__':
    sol = Solution()

    nums, target, expected = [-1,0,3,5,9,12], 9, 4
    assert sol.search(nums, target) == expected

    nums, target, expected = [-1,0,3,5,9,12], 2, -1
    assert sol.search(nums, target) == expected

    nums, target, expected = [-1,0,3,5,9,12], 13, -1
    assert sol.search(nums, target) == expected

File: 0081/binary_search.cpp
#include <gtest/gtest.h>

#include <cstdlib>
#include <vector>

using namespace std;

class Solution {
 public:
  bool search(vector<int>& nums, int target) {
    if (nums.size() == 0) { return false; }
    int left = 0, right = nums.size() - 1;
    while (left + 1 < right) {
      int mid = (right - left) / 2 + left;
      if (nums[mid] == target) { return true; }
      // the left half is sorted
      if (nums[left] < nums[mid]) {
        if (nums[left] <= target && target < nums[mid]) {
          right = mid;
        } else {
          left = mid;
        }
      // the right half is sorted
      } else if (nums[mid] < nums[right]) {
        if (nums[mid] < target && target <= nums[right]) {
          left = mid;
        } else {
          right = mid;
        }
      } else {
        if (nums[left] == nums[mid]) { left++; }
        if (nums[mid] == nums[right]) { right--; }
      }
    }
    return nums[left] == target || nums[right] == target;
  }
};

int main() {
  Solution sol;
  vector<int> nums;
  int target;

  nums = {2, 5, 6, 0, 0, 1, 2};
  target = 0;
  EXPECT_TRUE(sol.search(nums, target));
  target = 1;
  EXPECT_TRUE(sol.search(nums, target));
  target = 2;
  EXPECT_TRUE(sol.search(nums, target));
  target = 3;
  EXPECT_FALSE(sol.search(nums, target));
  target = 4;
  EXPECT_FALSE(sol.search(nums, target));
  target = 5;
  EXPECT_TRUE(sol.search(nums, target));
  target = 6;
  EXPECT_TRUE(sol.search(nums, target));
  target = 7;
  EXPECT_FALSE(sol.search(nums, target));

  nums = {}, target = 5;
  EXPECT_FALSE(sol.search(nums, target));

  exit(EXIT_SUCCESS);
}
